mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php

```diff
@@ -0,0 +1,73 @@
+<?php
+/**
+ *  Tests for str_starts_with()
+ *
+ * @package    spiralWebDb\centralHub\Tests\Unit\ConfigStore
+ * @since      1.3.0
+ * @author     Robert A. Gadon
+ * @link       https://github.com/rgadon107/cornerstone
+ * @license    GNU General Public License 2.0+
+ */
+
+namespace spiralWebDb\centralHub\Tests\Unit\ConfigStore;
+
+use function KnowTheCode\ConfigStore\str_starts_with;
+use spiralWebDb\Cornerstone\Tests\Unit\Test_Case;
+
+/**
+ * Class Tests_StringStartsWith
+ *
+ * @package spiralWebDb\centralHub\Tests\Unit\ConfigStore
+ * @group   config-store
+ */
+class Tests_StringStartsWith extends Test_Case {
+
+	/**
+	 * Prepares the test environment before each test.
+	 */
+	protected function setUp() {
+		parent::setUp();
+
+		require_once CENTRAL_HUB_ROOT_DIR . '/src/config-store/internals.php';
+	}
+
+	/*
+	 *  Test should return true when substring starts with and matches haystack.
+	 */
+	public function test_should_return_true_when_substring_starts_with() {
+		$this->assertTrue( str_starts_with( 'Hello World!', 'Hello' ) );
+		$this->assertTrue( str_starts_with( 'Hello World!', 'H' ) );
+		$this->assertTrue( str_starts_with( 'Cornerstone', 'Corner' ) );
+	}
+
+	/**
+	 * Test should return false when substring does not start with & match haystack.
+	 */
+	public function test_should_return_false_when_substring_does_not_start() {
+		$this->assertFalse( str_starts_with( 'Hello World', 'World' ) );
+		$this->assertFalse( str_starts_with( 'Hello World', 'hey!' ) );
+		$this->assertFalse( str_starts_with( 'Cornerstone', 'cor' ) );
+		$this->assertFalse( str_starts_with( 'Cornerstone', 'stone' ) );
+	}
+
+	/**
+	 * Test should return true when substring is empty string.
+	 */
+	public function test_should_return_true_when_substring_is_empty() {
+		$this->assertTrue( str_starts_with( 'Cornerstone', '' ) );
```

Comment:
@rgadon107 This one should not be true, as `Cornerstone` starts with a `C` and not an empty space.  When you run the unit tests, does this test pass or fail?  If it passes, then we need to look at the function to see why.

Reply:
@hellofromtonya I checked the documentation for the the PHP function [mb_substr](https://www.php.net/manual/en/function.mb-substr.php). The documentation says (in regards to the `$length` parameter): "If omitted or NULL is passed, extract all characters to the end of the string."

As you can see by the test, when I assign `$needle` an empty string in the function `str_starts_with()`, the function sets the variable `$needle_length` as the length of the entire string in `$haystack`. 

The API function `getAllKeysStartingWith( $starts_with )` will only work if it receives a valid string ( `$needle`) from `str_starts_with()`. I think that the function `str_starts_with()` should include a conditional to check for `$needle = ''`. If true, it should throw an Exception and message to inform the user that a string was expected and none was received. 

Does that make sense? Would you implement that approach?

Reply:
>If true, it should throw an Exception and message to inform the user that a string was expected and none was received.

Just a note:  An exception informs the developer.  We don't want exceptions to inform site owners or visitors to the site.  If you use an exception, then we hope that a developer will handle and fix the problem.

How would you validate that the function has received a haystack and needle to compare?  The lengths of both would need to be more than 0.

Consider the following:

```php
function str_starts_with( $haystack, $needle, $encoding = 'UTF-8' ) {
	$needle_length = mb_strlen( $needle, $encoding );

	// Using string lengths, check if given a haystack and needle to compare.
	if ( $needle_length == 0 || mb_strlen( $haystack, $encoding ) == 0 ) {
		throw new InvalidArgumentException(
			sprintf( 'The haystack and needle cannot be empty. Given: haystack [%s] and needle of [%s].', $haystack, $needle )
		);
	}

	return ( mb_substr( $haystack, 0, $needle_length, $encoding ) === $needle );
}
```

Notice that the code is using the length of the strings given.  If either string has 0 characters, then the error is thrown.

Why not use a boolean check for falsey, e.g. 

```php
if ( ! $haystack || ! $needle ) {
```
??  Why not?  Because a string of `'0'` would flag as a falsey and throw the error. Consider that you want to check if a string starts with a 0.  You couldn't with the falsey check.

Hint: Try it for yourself to validate what happens when you pass a `"0"` as either the haystack or needle with the falsey check:

```php
vdump_and_die( ( '0' ) );
```

Given that falsey isn't going to work when needing to check if it starts with 0, we need a different way to check.  We use the string's length, i.e. number of characters.  If there are no characters, that means an empty string was passed.